scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -211,7 +211,7 @@ class ScioContext private[scio] (val options: DataflowPipelineOptions,
       throw new RuntimeException("Cannot set name once pipeline is initialized")
     }
     // override app name and job name
-    options.setAppName(name)
+    if (options.getAppName == null) options.setAppName(name)
```

Comment:
Right. I'm not completely sure when the `@InstanceFactory(DataflowPipelineOptions.JobNameFactory.class)` thing is executed normally. Is it on each property read, or just after the underlying proxy is craeted (after parse).

I'll look into it and change to your suggestion if it works as expected.